example_allocators/compartment_alloc/README.md

```diff
@@ -0,0 +1,17 @@
+# Compartmentalising Allocator
+
+A modification of the "bump allocator" in this repo.
+This allocator creates compartments on the heap, and returns capabilities
```

Comment:
More of a philosophical question than asking for a specific change, but what does `compartments` mean in this instance?

Reply:
Great Q…

Here, the "compartment" is some heap-allocated space that only capabilities are handed out to. There are obviously ways to abuse "just some space on the heap" if the DDC isn't also restricted. The easiest way to explain this — I think — is to draw it out, so please excuse the long response…

I've begun to draw out compartments when I think about them like this. Assuming only the DDC is being used for compartmentalisation (so we're constraining, say, dereferencing of addresses for data but not jump points etc) and it's being used to bound which addresses can be legally dereferenced, we could draw memory out as a "number line" of the address space and mark points where a compartment's bounds begin and end. So we could diagram compartments like so:

```
|--------------------|
0  g   f      F   G  ♾️
```

where a compartment's lower bound is a lowercase letter and its upper bound the corresponding upper case letter (here ♾️ is the maximum address, and the `G` compartment's lower bound is marked by `g`, and the upper case marked by `G`). I'm obviously just defining compartment bounds relative to each other, and not in any kind of detail.

The compartments this example constructs are intended to sit _outwith_ all other compartments, so that any access performed against them _must_ be done through the capabilities returned by `malloc_compartment`. For this to work, obviously the DDC needs to be restricted when a program first runs, so that no future compartment's bounds will overlap with the ones we construct on the heap. Then, the ordinary compartments we create (derived from the now-restricted initial DDC) can't overlap with these separated heap-y compartments. I've taken to thinking of them as "side compartments" to differentiate them from ordinary compartments. To illustrate what I mean, we can create a special space adjacent to our ordinary compartments and allocations like this:

```
|---------------------|
0              ><     ♾️ 
```

Where `>` marks the upper bound of a DDC we install right when a program launches, and `<` marks the lower bound of possible "side" compartments. To extend the first diagram, we might make a side compartment that looks like…:

```
|------------------------------|
0  g   f    F   G  ><  p   P   ♾️
```

…and the "side" compartment `P` is what `compartment` means in this instance. Those compartments are similar to the buffers created in the `bump-allocator` example in structure, but this is long enough so I'll leave their definition in the code to speak to that. There's a little work to do, I suppose, depending on how thorough the example aims to be, but this is just a toy to demonstrate a kind of cheri-aware hybrid allocator.

Apologies for the wall of text, hope this is clear enough to explain what I mean :)

Reply:
(I've got a similar but slightly more detailled write-up of compartment structure and these "side" compartments I wondered about adding to [the notes on compartments that already exist](https://github.com/capablevms/cheri_compartments), in case this concept ends up being useful and the definition wants to live somewhere…)

Reply:
I've now added the above notes to the README in [da50830](https://github.com/capablevms/cheri-examples/pull/88/commits/da5083095393c676bf679fb4e5368fef8e4ae83c).